src/win/tty.c

```diff
@@ -937,30 +956,106 @@ int uv_tty_read_start(uv_tty_t* handle, uv_alloc_cb alloc_cb,
-    DWORD written;
     memset(&record, 0, sizeof record);
     if (!WriteConsoleInputW(handle->handle, &record, 1, &written)) {
       return GetLastError();
     }
+  } else if (!(handle->flags & UV_HANDLE_CANCELLATION_PENDING)) {
+    /* Cancel line-buffered read if not already pending */
+    err = uv__cancel_read_console(handle);
+    if (err)
+      return err;
+
+    handle->flags |= UV_HANDLE_CANCELLATION_PENDING;
+  }
+
+  return 0;
+}
+
+static int uv__cancel_read_console(uv_tty_t* handle) {
+  HANDLE active_screen_buffer = INVALID_HANDLE_VALUE;
+  CONSOLE_SCREEN_BUFFER_INFO info;
+  BOOL restore_screen_state;
+  INPUT_RECORD record;
+  DWORD written;
+  DWORD err = 0;
+
+  EnterCriticalSection(&uv__read_console_trap_lock);
+  if (!uv__read_console_started) {
+    /* We have managed to trap the other thread before ReadConsole is called */
+    uv__read_console_trap = TRUE;
+    LeaveCriticalSection(&uv__read_console_trap_lock);
+    return 0;
   }
+  LeaveCriticalSection(&uv__read_console_trap_lock);
+
+  /* If ReadConsole already returned (the user pressed ENTER),
+     there is nothing to do */
+  if (WaitForSingleObject(uv__read_console_done_event, 0) == WAIT_OBJECT_0)
+    return 0;
+
+  /* From this point return should only happen by jumping to the out label
+     to ensure proper cleanup */
+
+  /* Save screen state before sending the VK_RETURN event */
+  active_screen_buffer = CreateFileA("conout$",
+                                     GENERIC_READ | GENERIC_WRITE,
+                                     FILE_SHARE_READ | FILE_SHARE_WRITE,
+                                     NULL,
+                                     OPEN_EXISTING,
+                                     FILE_ATTRIBUTE_NORMAL,
+                                     NULL);
 
-  /* Cancel line-buffered read */
-  if (handle->tty.rd.read_line_handle != NULL) {
-    /* Closing this handle will cancel the ReadConsole operation */
-    CloseHandle(handle->tty.rd.read_line_handle);
-    handle->tty.rd.read_line_handle = NULL;
+  if (active_screen_buffer != INVALID_HANDLE_VALUE &&
+      GetConsoleScreenBufferInfo(active_screen_buffer, &info)) {
+    restore_screen_state = TRUE;
+  } else {
+    restore_screen_state = FALSE;
+  }
+
+  /* Write enter key event to force the console wait to return. */
+  record.EventType = KEY_EVENT;
+  record.Event.KeyEvent.bKeyDown = TRUE;
+  record.Event.KeyEvent.wRepeatCount = 1;
+  record.Event.KeyEvent.wVirtualKeyCode = VK_RETURN;
+  record.Event.KeyEvent.wVirtualScanCode =
+    MapVirtualKeyW(VK_RETURN, MAPVK_VK_TO_VSC);
+  record.Event.KeyEvent.uChar.UnicodeChar = L'\r';
+  record.Event.KeyEvent.dwControlKeyState = 0;
+  if (!WriteConsoleInputW(handle->handle, &record, 1, &written)) {
+    err = GetLastError();
+    goto out;
+  } else {
+    /* At this point, we have sent the VK_RETURN.
+       Any further errors shouldn't be be returned. */
+    err = 0;
   }
 
+  WaitForSingleObject(uv__read_console_done_event, INFINITE);
```

Comment:
I agree it would be preferable not to wait. We were a bit concerned about what could happen with all the possible read_stop/read_start/set_mode sequences, and this way also minimizes the time between cursors jumps, but assuming nothing can modify the screen buffer state while UV_HANDLE_READ_PENDING is set, it should be doable. I'll take another look.